src/dataset.jl

```diff
@@ -1,78 +1,118 @@
+that implements the same interface as `DimensionalData.DimStack` and has identical usage.
 
-  - `coords::Dict{String,Any}`: Dict mapping dimension names to index names. Possible
-    arguments has same form as `data_vars`.
-  - `attrs::Dict{String,Any}`: Global attributes to save on this dataset.
+When a `Dataset` is passed to Python, it is converted to an `xarray.Dataset` without copying
+the data. That is, the Python object shares the same memory as the Julia object. However,
+if an `xarray.Dataset` is passed to Julia, its data must be copied.
 
 In most cases, use [`convert_to_dataset`](@ref) or [`convert_to_constant_dataset`](@ref) or
 to create a `Dataset` instead of directly using a constructor.
 """
-struct Dataset
-    o::PyObject
-
-    function Dataset(o::PyObject)
-        if !pyisinstance(o, xarray.Dataset)
-            throw(ArgumentError("$o is not an `xarray.Dataset`."))
-        end
-        return new(o)
-    end
+struct Dataset{L,D<:Tuple,R<:Tuple,LD<:NamedTuple,M,LM<:NamedTuple} <:
+       DimensionalData.AbstractDimStack{L}
+    data::L
+    dims::D
+    refdims::R
+    layerdims::LD
+    metadata::M
+    layermetadata::LM
+end
+
+Dataset(das::DimensionalData.AbstractDimArray...; kw...) = Dataset(das; kw...)
+function Dataset(das::Tuple{Vararg{<:DimensionalData.AbstractDimArray}}; kw...)
+    return Dataset(NamedTuple{DimensionalData.uniquekeys(das)}(das); kw...)
+end
+function Dataset(
+    das::NamedTuple{<:Any,<:Tuple{Vararg{<:DimensionalData.AbstractDimArray}}};
+    data=map(parent, das),
+    dims=DimensionalData.combinedims(das...),
+    layerdims=map(DimensionalData.basedims, das),
+    refdims=(),
+    metadata=DimensionalData.NoMetadata(),
+    layermetadata=map(DimensionalData.metadata, das),
+)
+    return Dataset(data, dims, refdims, layerdims, metadata, layermetadata)
+end
+# Same sized arrays
+function Dataset(
+    data::NamedTuple,
+    dims::Tuple;
+    refdims=(),
+    metadata=DimensionalData.NoMetadata(),
+    layermetadata=map(_ -> DimensionalData.NoMetadata(), data),
+)
+    all(map(d -> axes(d) == axes(first(data)), data)) || throw(
+        ArgumentError(
+            "Arrays must have identical axes. For mixed dimensions, use DimArrays`"
+        ),
+    )
+    layerdims = map(_ -> DimensionalData.basedims(dims), data)
+    return Dataset(
+        data,
+        DimensionalData.format(dims, first(data)),
+        refdims,
+        layerdims,
+        metadata,
+        layermetadata,
+    )
 end
+Dataset(data::Dataset) = data
 
-Dataset(; kwargs...) = Dataset(xarray.Dataset(; kwargs...))
-@inline Dataset(data::Dataset) = data
+Base.parent(data::Dataset) = getfield(data, :data)
 
-@inline PyObject(data::Dataset) = getfield(data, :o)
+PyObject(data::Dataset) = _to_xarray(data)
 
-Base.convert(::Type{Dataset}, obj::PyObject) = Dataset(obj)
+Base.convert(::Type{Dataset}, obj::PyObject) = _dataset_from_xarray(obj)
```

Comment:
This is piratical in that if a user already loaded xarray using PyCall to use it, then if ArviZ comes into scope, their `xarray.Dataset`s will all be automatically converted to `ArviZ.Dataset`s. It might be pretty hard though to not have this automatic converter and instead add manual conversions everywhere where necessary.